File: sdks/javascript/src/resources/webhooks.ts
/**
 * Webhooks resource for managing webhook endpoints
 */

import type { Mailat } from '../client'
import type {
  Webhook,
  CreateWebhookOptions,
  UpdateWebhookOptions,
  WebhookCall,
  WebhookEvent
} from '../types'

export class WebhooksResource {
  constructor(private client: Mailat) {}

  /**
   * Create a new webhook
   *
   * @example
   * ```typescript
   * const webhook = await client.webhooks.create({
   *   name: 'My Webhook',
   *   url: 'https://example.com/webhook',
   *   events: ['email.delivered', 'email.bounced']
   * })
   * ```
   */
  async create(options: CreateWebhookOptions): Promise<Webhook> {
    return this.client.post<Webhook>('/webhooks', options)
  }

  /**
   * Get webhook by ID
   */
  async get(id: string): Promise<Webhook> {
    return this.client.get<Webhook>(`/webhooks/${id}`)
  }

  /**
   * Update a webhook
   */
  async update(id: string, options: UpdateWebhookOptions): Promise<Webhook> {
    return this.client.put<Webhook>(`/webhooks/${id}`, options)
  }

  /**
   * Delete a webhook
   */
  async delete(id: string): Promise<void> {
    await this.client.delete(`/webhooks/${id}`)
  }

  /**
   * List all webhooks
   */
  async list(): Promise<Webhook[]> {
    return this.client.get<Webhook[]>('/webhooks')
  }

  /**
   * Enable a webhook
   */
  async enable(id: string): Promise<Webhook> {
    return this.update(id, { active: true })
  }

  /**
   * Disable a webhook
   */
  async disable(id: string): Promise<Webhook> {
    return this.update(id, { active: false })
  }

  /**
   * Rotate webhook secret
   * Returns the new secret (shown only once)
   */
  async rotateSecret(id: string): Promise<{ secret: string }> {
    return this.client.post(`/webhooks/${id}/rotate-secret`)
  }

  /**
   * Test a webhook by sending a test payload
   */
  async test(id: string): Promise<{
    success: boolean
    statusCode?: number
    responseTime?: number
    error?: string
  }> {
    return this.client.post(`/webhooks/${id}/test`)
  }

  /**
   * Get recent webhook calls/deliveries
   */
  async getCalls(id: string, options: {
    page?: number
    limit?: number
    status?: 'success' | 'failed' | 'pending'
  } = {}): Promise<WebhookCall[]> {
    return this.client.get<WebhookCall[]>(`/webhooks/${id}/calls`, {
      page: options.page,
      limit: options.limit,
      status: options.status
    })
  }

  /**
   * Get available webhook event types
   */
  getEventTypes(): WebhookEvent[] {
    return [
      'email.sent',
      'email.delivered',
      'email.opened',
      'email.clicked',
      'email.bounced',
      'email.complained',
      'contact.created',
      'contact.updated',
      'contact.unsubscribed',
      'campaign.sent',
      'campaign.completed'
    ]
  }

  /**
   * Verify webhook signature (for receiving webhooks)
   *
   * @example
   * ```typescript
   * const isValid = client.webhooks.verifySignature(
   *   requestBody,
   *   request.headers['x-webhook-signature'],
   *   webhookSecret
   * )
   * ```
   */
  verifySignature(payload: string | Buffer, signature: string, secret: string): boolean {
    // Use Web Crypto API or Node.js crypto
    if (typeof globalThis.crypto !== 'undefined' && globalThis.crypto.subtle) {
      // Browser environment - signature verification should be done server-side
      console.warn('Webhook signature verification should be done server-side')
      return false
    }

    // Node.js environment
    try {
      const crypto = require('crypto')
      const expectedSignature = crypto
        .createHmac('sha256', secret)
        .update(payload)
        .digest('hex')

      const providedSignature = signature.replace('sha256=', '')
      return crypto.timingSafeEqual(
        Buffer.from(expectedSignature),
        Buffer.from(providedSignature)
      )
    } catch {
      return false
    }
  }
}

File: sdks/python/mailat/resources/webhooks.py
"""Webhooks resource for managing webhook endpoints"""

from typing import TYPE_CHECKING, Optional
import hmac
import hashlib
from ..types import Webhook, WebhookEvent, WebhookCall

if TYPE_CHECKING:
    from ..client import Mailat


class WebhooksResource:
    """Resource for managing webhook endpoints."""

    def __init__(self, client: "Mailat"):
        self._client = client

    def create(
        self,
        name: str,
        url: str,
        events: list[WebhookEvent | str],
    ) -> Webhook:
        """Create a new webhook."""
        event_list = [e.value if isinstance(e, WebhookEvent) else e for e in events]
        data = self._client.post("/webhooks", json={
            "name": name,
            "url": url,
            "events": event_list,
        })
        return Webhook(**data)

    def get(self, webhook_id: str) -> Webhook:
        """Get webhook by ID."""
        data = self._client.get(f"/webhooks/{webhook_id}")
        return Webhook(**data)

    def update(
        self,
        webhook_id: str,
        name: Optional[str] = None,
        url: Optional[str] = None,
        events: Optional[list[WebhookEvent | str]] = None,
        active: Optional[bool] = None,
    ) -> Webhook:
        """Update a webhook."""
        payload = {}
        if name:
            payload["name"] = name
        if url:
            payload["url"] = url
        if events:
            payload["events"] = [e.value if isinstance(e, WebhookEvent) else e for e in events]
        if active is not None:
            payload["active"] = active

        data = self._client.put(f"/webhooks/{webhook_id}", json=payload)
        return Webhook(**data)

    def delete(self, webhook_id: str) -> None:
        """Delete a webhook."""
        self._client.delete(f"/webhooks/{webhook_id}")

    def list(self) -> list[Webhook]:
        """List all webhooks."""
        data = self._client.get("/webhooks")
        return [Webhook(**w) for w in data]

    def enable(self, webhook_id: str) -> Webhook:
        """Enable a webhook."""
        return self.update(webhook_id, active=True)

    def disable(self, webhook_id: str) -> Webhook:
        """Disable a webhook."""
        return self.update(webhook_id, active=False)

    def rotate_secret(self, webhook_id: str) -> dict[str, str]:
        """Rotate webhook secret."""
        return self._client.post(f"/webhooks/{webhook_id}/rotate-secret")

    def test(self, webhook_id: str) -> dict:
        """Test a webhook."""
        return self._client.post(f"/webhooks/{webhook_id}/test")

    def get_calls(
        self,
        webhook_id: str,
        page: int = 1,
        limit: int = 50,
        status: Optional[str] = None,
    ) -> list[WebhookCall]:
        """Get recent webhook calls."""
        params = {"page": page, "limit": limit}
        if status:
            params["status"] = status

        data = self._client.get(f"/webhooks/{webhook_id}/calls", params=params)
        return [WebhookCall(**c) for c in data]

    @staticmethod
    def get_event_types() -> list[str]:
        """Get available webhook event types."""
        return [e.value for e in WebhookEvent]

    @staticmethod
    def verify_signature(payload: str | bytes, signature: str, secret: str) -> bool:
        """Verify webhook signature."""
        if isinstance(payload, str):
            payload = payload.encode()

        expected = hmac.new(
            secret.encode(),
            payload,
            hashlib.sha256
        ).hexdigest()

        provided = signature.replace("sha256=", "")
        return hmac.compare_digest(expected, provided)
